src/main/scala/org/apache/spark/shuffle/ucx/UcxShuffleTransport.scala

```diff
@@ -0,0 +1,288 @@
+    clientConnections.values.foreach(ep => ep.close())
+    registeredBlocks.forEachKey(1, blockId => unregister(blockId))
+    allocatedWorkers.forEach(_.close())
+    globalWorker.close()
+    ucxContext.close()
+  }
+
+  /**
+   * Add executor's worker address. For standalone testing purpose and for implementations that makes
+   * connection establishment outside of UcxShuffleManager.
+   */
+  def addExecutor(executorId: String, workerAddress: ByteBuffer): Unit = {
+    executorIdToAddress.put(executorId, workerAddress)
+  }
+
+  /**
+   * On a sender side process request of fetchBlockByBlockId
+   */
+  private[ucx] def replyFetchBlockRequest(memoryBlock: MemoryBlock) {
+    val buffer = UcxUtils.getByteBufferView(memoryBlock.address, memoryBlock.size.toInt)
+
+    val tag = buffer.getLong()
+    val idSize = buffer.getInt()
+    val idArray = new Array[Byte](idSize)
+    buffer.get(idArray)
+
+    val id = new String(idArray)
+    val workerAddressSize = buffer.getInt()
+    val workerAddress = buffer.slice()
+
+    workerAddress.limit(workerAddressSize)
+    buffer.position(buffer.position() + workerAddressSize)
+
+    val ep = clientConnections.getOrElseUpdate(id,
+      globalWorker.newEndpoint(new UcpEndpointParams().setUcpAddress(workerAddress))
+    )
+
+    val blockId: BlockId =
+      Utils.tryWithResource(new ByteBufferBackedInputStream(buffer))(bin => {
+        val in = new ObjectInputStream(bin)
+        in.readObject().asInstanceOf[BlockId]
+    })
+
+    val block = registeredBlocks.get(blockId)
+    if (block == null) {
+      throw new UcxException(s"Block $blockId not registered")
+    }
+    val lock = block.lock.readLock()
+    lock.lock()
+    val blockMemory = block.getMemoryBlock
+
+    ep.sendTaggedNonBlocking(blockMemory.address, blockMemory.size, tag, new UcxCallback {
+      override def onSuccess(request: UcpRequest): Unit = {
+        lock.unlock()
+      }
+
+      override def onError(ucsStatus: Int, errorMsg: String): Unit = {
+        logError(s"Failed to send $blockId: $errorMsg")
+        lock.unlock()
+      }
+    })
+  }
+
+  private def pinMemory(block: Block): UcpMemory = {
+    val blockMemory = block.getMemoryBlock
+    ucxContext.memoryMap(
+      memMapParams.setAddress(blockMemory.address).setLength(blockMemory.size))
+  }
+
+  /**
+   * Registers blocks using blockId on SERVER side.
+   */
+  override def register(blockId: BlockId, block: Block) {
+    val registeredBock: Block = if (ucxShuffleConf.pinMemory) {
+      UcxPinnedBlock(block, pinMemory(block))
+    } else {
+      block
+    }
+    registeredBlocks.put(blockId, registeredBock)
+  }
+
+  /**
+   * Change location of underlying blockId in memory
+   */
+  override def mutate(blockId: BlockId, block: Block, callback: OperationCallback): Unit = {
+    Future {
+      unregister(blockId)
+      register(blockId, block)
+    } andThen {
+      case Failure(t) => if (callback != null) {
+        callback.onComplete(new OperationResult {
```

Comment:
could these two callbacks leverage `UcxSuccessOperationResult` and `UcxFailureOperationResult`?